Add half-precision support for `DepthwiseConv2dNative`

yongtang opened the pull request:
This fix tries to address the request raised in #12327 so that float16 is supported for `DepthwiseConv2dNative`

This fix fixes #12327.

This fix fixes #11502.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add float16 support for `DepthwiseConv2dNative`

This fix tries to address the request raised in 12327
so that float16 is supported for `DepthwiseConv2dNative`

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add test cases for float16 support of `DepthwiseConv2dNative`

A couple of tests for float16 have been disabled as error is
large.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

tensorflow-jenkins commented:
Can one of the admins verify this patch?

drpngx commented on tensorflow/core/kernels/depthwise_conv_grad_op.cc:
```diff
@@ -231,7 +231,7 @@ static void CopyOutputBackpropRegion(const DepthwiseArgs& args,
       // Pad to vector-register width (if needed).
       for (int64 d = 0; d < pad_size; ++d) {
-        buffer[buf_base + vectorized_size + scalar_size + d] = 0;
+        buffer[buf_base + vectorized_size + scalar_size + d] = T(0);
```
Can we use `static_cast<T>(0)` instead?

yongtang replied:
Thanks @drpngx, done with the change `static_cast<T>(0)` and other similar places.

drpngx commented on tensorflow/python/framework/test_util.py:
```diff
@@ -752,7 +752,8 @@ def assertArrayNear(self, farray1, farray2, err):
     self.assertEqual(len(farray1), len(farray2))
     for f1, f2 in zip(farray1, farray2):
-      self.assertNear(float(f1), float(f2), err)
+      if float(f1) != float(f2):
```
Why this change?

yongtang replied:
@drpngx The purpose for checking `f1 == f2` is for the case where `f1, f2 = inf, inf`. `assertNear` used `math.abs(f1-f2)` which will not cover `inf, inf` case. 

I have moved the `f1 == f2` checking to `assertNear` with comments added, so that it is easy to see the intention.

drpngx commented on tensorflow/python/kernel_tests/depthwise_conv_op_test.py:
```diff
@@ -369,6 +373,7 @@ def testDepthwiseConv2DInputGrad(self):
       print("Testing DepthwiseConv2DInputGrad,", index, "th config:",
             input_size, "*", filter_size, "stride:", stride, "padding:",
             padding)
+      # float16 test is disabled
```
Please punctuate comments `The test for float16 is disabled.` I would be in favor of removing that comment however.

Is it really disabled?

yongtang replied:
@drpngx The comments has been updated.

yongtang commented:
@drpngx Thanks for the review. The PR has been updated. Please take a look and let me know if there are any issues.

drpngx commented:
Jenkins, test this please.

drpngx commented:
Sorry to be dense, but where is the float16 test disabled, and can you report what the differences are?

drpngx commented:
Strange, it seems to fail on Mac on `import enum`.

yongtang commented:
@drpngx Thanks for the review.

The PR didn't `disable` the float16, but rather `not enable` the float16 tests for `testDepthwiseConv2DInputGrad` and `testDepthwiseConv2DFilterGrad`.

Before this PR, in `testDepthwiseConv2DInputGrad` (Ln 371):
```
      for data_type in [dtypes.float32, dtypes.float64]:
        self._ConstructAndTestGradient(
            input_size,
            filter_size,
            output_size,
            stride,
            padding,
            data_type,
            test_input=True,
            use_gpu=True)
```

After this PR, the `dtypes.float16` was **not added** to the above. Instead, a comment was added to specify that `dtypes.float16` is not enabled (yet).

If we enable `dtypes.float16`, then here is the error message:
```
======================================================================
FAIL: testDepthwiseConv2DInputGrad (__main__.DepthwiseConv2DTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/ubuntu/.cache/bazel/_bazel_ubuntu/ad1e09741bb4109fbc70ef8216b59ee2/execroot/org_tensorflow/bazel-out/local-opt/bin/tensorflow/python/kernel_tests/depthwise_conv_op_test.runfiles/org_tensorflow/tensorflow/python/kernel_tests/depthwise_conv_op_test.py", line 387, in testDepthwiseConv2DInputGrad
    use_gpu=True)
  File "/home/ubuntu/.cache/bazel/_bazel_ubuntu/ad1e09741bb4109fbc70ef8216b59ee2/execroot/org_tensorflow/bazel-out/local-opt/bin/tensorflow/python/kernel_tests/depthwise_conv_op_test.runfiles/org_tensorflow/tensorflow/python/kernel_tests/depthwise_conv_op_test.py", line 368, in _ConstructAndTestGradient
    self.assertLess(err, tolerance)
AssertionError: 3.373046875 not less than 0.002

----------------------------------------------------------------------
Ran 10 tests in 21.389s

FAILED (failures=1)
```

Please take a look.

sb2nov commented:
@yongtang a lot of the tests seemed to fail with `ImportError: No module named enum` can you address that to keep the review moving forward.

```
Traceback (most recent call last):
  File "/private/var/tmp/_bazel_jenkins/13e370a18c169b19baeafefb05212b85/execroot/tensorflow-pull-requests-mac/bazel-out/darwin_x86_64-opt/bin/tensorflow/python/debug/examples_test.runfiles/org_tensorflow/tensorflow/python/debug/examples/debug_fibonacci.py", line 25, in <module>
    import tensorflow as tf
  File "/private/var/tmp/_bazel_jenkins/13e370a18c169b19baeafefb05212b85/execroot/tensorflow-pull-requests-mac/bazel-out/darwin_x86_64-opt/bin/tensorflow/python/debug/examples_test.runfiles/org_tensorflow/tensorflow/__init__.py", line 24, in <module>
    from tensorflow.python import *
  File "/private/var/tmp/_bazel_jenkins/13e370a18c169b19baeafefb05212b85/execroot/tensorflow-pull-requests-mac/bazel-out/darwin_x86_64-opt/bin/tensorflow/python/debug/examples_test.runfiles/org_tensorflow/tensorflow/python/__init__.py", line 63, in <module>
    from tensorflow.python.framework.framework_lib import *
  File "/private/var/tmp/_bazel_jenkins/13e370a18c169b19baeafefb05212b85/execroot/tensorflow-pull-requests-mac/bazel-out/darwin_x86_64-opt/bin/tensorflow/python/debug/examples_test.runfiles/org_tensorflow/tensorflow/python/framework/framework_lib.py", line 102, in <module>
    from tensorflow.python.framework.importer import import_graph_def
  File "/private/var/tmp/_bazel_jenkins/13e370a18c169b19baeafefb05212b85/execroot/tensorflow-pull-requests-mac/bazel-out/darwin_x86_64-opt/bin/tensorflow/python/debug/examples_test.runfiles/org_tensorflow/tensorflow/python/framework/importer.py", line 30, in <module>
    from tensorflow.python.framework import function
  File "/private/var/tmp/_bazel_jenkins/13e370a18c169b19baeafefb05212b85/execroot/tensorflow-pull-requests-mac/bazel-out/darwin_x86_64-opt/bin/tensorflow/python/debug/examples_test.runfiles/org_tensorflow/tensorflow/python/framework/function.py", line 36, in <module>
    from tensorflow.python.ops import variable_scope as vs
  File "/private/var/tmp/_bazel_jenkins/13e370a18c169b19baeafefb05212b85/execroot/tensorflow-pull-requests-mac/bazel-out/darwin_x86_64-opt/bin/tensorflow/python/debug/examples_test.runfiles/org_tensorflow/tensorflow/python/ops/variable_scope.py", line 24, in <module>
    import enum  # pylint: disable=g-bad-import-order
ImportError: No module named enum
```

sb2nov commented:
As it is not your change, you might just have to rebase.

yongtang committed:
Add float16 support for `DepthwiseConv2dNative` on GPU

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

Add Eigen::half support for DepthwiseConv2dNative on GPU

This commit add Eigen::half support for DepthwiseConv2dNative on GPU

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @sb2nov. I rebased the PR and also fixed several compilation errors on GPU. The tests should pass now. Please take a look.

sb2nov commented:
Jenkins, test this please.

yongtang commented:
The Jenkins failure for `Linux CPU Test Makefile` seems to be related to permissions on Jenkins server?
```
rm -rf /workspace/tensorflow/contrib/makefile/gen
rm: cannot remove '/workspace/tensorflow/contrib/makefile/gen/obj/tensorflow/core/graph/mkl_tfconversion_pass.o': Permission denied
rm: cannot remove '/workspace/tensorflow/contrib/makefile/gen/obj/tensorflow/core/graph/graph_partition.o': Permission denied
rm: cannot remove '/workspace/tensorflow/contrib/makefile/gen/obj/tensorflow/core/graph/mkl_layout_pass.o': Permission denied
```

drpngx commented:
Transient error. Jenkins, test this please

On Sep 26, 2017 4:43 AM, "Yong Tang" <notifications@github.com> wrote:

> The Jenkins failure for Linux CPU Test Makefile seems to be related to
> permissions on Jenkins server?
>
> rm -rf /workspace/tensorflow/contrib/makefile/gen
> rm: cannot remove '/workspace/tensorflow/contrib/makefile/gen/obj/tensorflow/core/graph/mkl_tfconversion_pass.o': Permission denied
> rm: cannot remove '/workspace/tensorflow/contrib/makefile/gen/obj/tensorflow/core/graph/graph_partition.o': Permission denied
> rm: cannot remove '/workspace/tensorflow/contrib/makefile/gen/obj/tensorflow/core/graph/mkl_layout_pass.o': Permission denied
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/12976#issuecomment-332171579>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AT_SbXwI996ILYdw9ZOR3xA_MaYyRhU9ks5smONMgaJpZM4PTjjk>
> .
>

drpngx commented:
How imprecise is the `float16` result?

yongtang commented:
@drpngx The `depthwise_conv2d_native` op itself is fine. The imprecise is for the `gradient` and the error is pretty large:


```
======================================================================
FAIL: testDepthwiseConv2DFilterGrad (__main__.DepthwiseConv2DTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/ubuntu/.cache/bazel/_bazel_ubuntu/ad1e09741bb4109fbc70ef8216b59ee2/execroot/org_tensorflow/bazel-out/local-opt/bin/tensorflow/python/kernel_tests/depthwise_conv_op_test.runfiles/org_tensorflow/tensorflow/python/kernel_tests/depthwise_conv_op_test.py", line 427, in testDepthwiseConv2DFilterGrad
    use_gpu=True)
  File "/home/ubuntu/.cache/bazel/_bazel_ubuntu/ad1e09741bb4109fbc70ef8216b59ee2/execroot/org_tensorflow/bazel-out/local-opt/bin/tensorflow/python/kernel_tests/depthwise_conv_op_test.runfiles/org_tensorflow/tensorflow/python/kernel_tests/depthwise_conv_op_test.py", line 368, in _ConstructAndTestGradient
    self.assertLess(err, tolerance)
AssertionError: 3.29296875 not less than 0.002

======================================================================
FAIL: testDepthwiseConv2DInputGrad (__main__.DepthwiseConv2DTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/ubuntu/.cache/bazel/_bazel_ubuntu/ad1e09741bb4109fbc70ef8216b59ee2/execroot/org_tensorflow/bazel-out/local-opt/bin/tensorflow/python/kernel_tests/depthwise_conv_op_test.runfiles/org_tensorflow/tensorflow/python/kernel_tests/depthwise_conv_op_test.py", line 387, in testDepthwiseConv2DInputGrad
    use_gpu=True)
  File "/home/ubuntu/.cache/bazel/_bazel_ubuntu/ad1e09741bb4109fbc70ef8216b59ee2/execroot/org_tensorflow/bazel-out/local-opt/bin/tensorflow/python/kernel_tests/depthwise_conv_op_test.runfiles/org_tensorflow/tensorflow/python/kernel_tests/depthwise_conv_op_test.py", line 368, in _ConstructAndTestGradient
    self.assertLess(err, tolerance)
AssertionError: 3.373046875 not less than 0.002

----------------------------------------------------------------------
```

drpngx commented:
That seems a little too large. Maybe it's a bug? I'm afraid that people will get confused if they try to train with it. Can we just disable it for float16?

yongtang committed:
Disable float16 support for the gradient of DepthwiseConv2dNative

The accuracy is not high enough for float16 of gradient.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
@drpngx The gradient of `DepthwiseConv2dNative` for float16 has been disabled. Please take a look.

martinwicke commented:
Jenkins, test this please.

martinwicke reviewed (approved):
Approving; @drpngx's comments have been addressed.

martinwicke commented:
Jenkins, test this please.

I should have fixed the MacOS and XLA failures.
